### DISCUSSION TITLE
V2 4.2.2 BLT firmware not using mesh with recommended start gcode?

### DISCUSSION THREAD

**Discussion by cr08** [mriscoc/Ender3V2S1#264] [Q&A] [answered]
Currently working through some print issues on my E3 V2. Latest thing I'm seeing and not 100% sure if this is intentional but doesn't seem to be as I'm not even seeing any Z movement but while printing the first layer after a mesh has been completed and saved, the Z level on the screen is showing zero movement across the entire bed.

Using the latest Cura 5.1.0 with the start gcode as it is listed in the Wiki here. As I've been trying to solve some other print issues, I've basically been running the Tramming wizard and building a new 7x7 mesh before each print. But seems I'm not having any luck getting it to actually use the mesh in a print. And this is evident using a bed level test print with the four corners and center not having consistent lines and pressure.

For reference, here's my full start gcode:
````
M201 X500.00 Y500.00 Z100.00 E5000.00 ;Setup machine max acceleration
M203 X500.00 Y500.00 Z20.00 E50.00 ;Setup machine max feedrate
M204 P500.00 R1000.00 T500.00 ;Setup Print/Retract/Travel acceleration
M205 X8.00 Y8.00 Z0.40 E5.00 ;Setup Jerk
M220 S100 ;Reset Feedrate
M221 S100 ;Reset Flowrate

M190 S60 ; preheat bed for abl
G92 E0 ; Reset Extruder
G28 ; Home all axes
M420 S1 Z2; Use Level Mesh up to 2mm;
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed

G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position
G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
G92 E0 ; Reset Extruder
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish
````

**Comment by mriscoc** [comment]
Mesh leveling compensation does not affect the Z value in the screen. Check if you have a valid mesh by open the mesh viewer.

**Reply by DoogH2004** [reply]
I'm having this same issue.  My mesh seems to be ignored.  I have "activate leveling" selected before starting a print, and then after starting a print and returning to settings, it will be turned back off.  This also seems to be reflected in the areas of the print bed that are not sticking.  For example, the bottom right of the board is offset by about -0.5mm, but if I watch the print head, it does not adjust for this mesh correction.  It's pretty easy to eyeball, the print head develops a visible gap between the base and hot end.

I am glad I am not the only one!

And yes, I have tried both including (and excluding) M420 S1 in the gcode.

**Comment by mriscoc** [comment, marked as answer]
"Activate leveling" will be always set back to off after any G28, it is re-enabled by M420 S1 (or G29 A for UBL versions). Currently there isn't any problem with leveling in Ender V2 with v4.2.2 boards firmware.

**Reply by ghost** [reply]
With me it is working wonderfully with my Ender 3 V2, it has 3dtouch installed.
The mesh I use is 5x5
I use CURA 5.1 and in gcode I am using the G29 command just below the G28.
My board is a 4.2.2 version

**Reply by DoogH2004** [reply]
I'm using the start gcode from the wiki (which has M420 after G28), but it is having this issue.  I tried swapping to UBL instead, and the issue seems to be resolved.

Not sure why.

**Reply by DoogH2004** [reply]
Edit:
@mriscoc 
The issue is not resolved.

For some reason, Mesh leveling is still always turning off, even after switching to UBL.

I have done two things in an attempt to fix this:

1.) I uncommented #define RESTORE_LEVELING_AFTER_G28

2.) I tried making sure my GCode always enables Auto-Leveling.  See below.

However, this will always run another auto-level sequence, without activating auto-level.  I have no idea why.

```
M201 X500.00 Y500.00 Z100.00 E5000.00 ;Setup machine max acceleration
M203 X500.00 Y500.00 Z20.00 E50.00 ;Setup machine max feedrate
M204 P500.00 R1000.00 T500.00 ;Setup Print/Retract/Travel acceleration
M205 X8.00 Y8.00 Z0.40 E5.00 ;Setup Jerk
M220 S100 ;Reset Feedrate
M221 S100 ;Reset Flowrate

G92 E0 ; Reset Extruder
G28 ; Home all axes
G29 A;      Bed-leveling on
G29 L0;     I have also tried L1
G29 J;        3-point tilt test
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5.1 Y20 Z0.3 F5000.0 ; Move to start position
G1 X5.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
G1 X5.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
G1 X5.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
G92 E0 ; Reset Extruder
G1 Z2.0 F3000 ; Move Z Axis up little to prevent scratching of Heat Bed
G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish
```

**Reply by MarkforU** [reply]
I'm running UBL, after I do a bed leveling 9x9 and save it to slot 0 or 1 or both. I leave that menu and save all settings. I can turn my machine off, and when I turn back on the first thing that pops up on screen is Mesh Loaded 0 or 1 depending on which slot I saved leaving it on after the save and power off. My Gcode is very simple and does UBL probe before print and everything is perfect. I'm running a V2 with the 4.2.2 board and a upgraded Ender 3 with new display with 4.2.7 board both with the latest UBL Firmware...I can also go to advanced menu under leveling and my box is still checked for Auto Bed Level Enabled from my last save and shutdown. Are you ticking that box and leaving the menu and save all machine settings?

**Reply by DoogH2004** [reply]
Yes, I am doing that.  I think I have a lead.  I'll report back.

**Comment by kileywm** [comment]
I'm experiencing the same issue as cr08 and DoogH2004 on an Ender 3 S1 (F1). It seems like my mesh isn't being applied.

Here is how my testing has gone:
1. Load Ender3S1-F1-UBL-20220716.bin firmware (also tested non-UBL firmware)
2. Tram the bed until tolerance is met
3. Calibrate z-offset
4. Generate 9x9 mesh (tried with multi-step set to 0 and 2, no change in results)
5. Save mesh to slot 0 and slot 1
6. View mesh to confirm valid mesh is generated
8. Toggle automatic bed leveling on/off and view the mesh compensation GUI indicator changing color correctly to show that status
9. Power off, then on, then confirm that saved mesh was correctly persisted and loaded on boot
10. Load https://www.thingiverse.com/thing:2789086 and slice with PrusaSlicer
11. Try both `M420 S1 Z2`, `M420 S1 L0 Z2` , and `G29 A` (separate test runs) to activate bed leveling immediately after `G28` command
12. Start print and observe the GUI indicator for mesh compensation turning off during `G28` and back on during `G29 A` and same result for `M420 S1 ...` commands in separate tests
13. Print results in no correction applied. Filament clearly squished at high spots, gapped at low spots, and good where the bed was naturally level

I also tested swapping steps 3 & 4 to see if the z-offset value was negatively affecting the mesh generation and saw no discernible difference.

Here is my starting g-code for reference
```
G90 ; use absolute coordinates
M83 ; extruder relative mode
M140 S{first_layer_bed_temperature[0]} ; set final bed temp
M104 S150 ; set temporary nozzle temp to prevent oozing during homing
G4 S10 ; allow partial nozzle warmup
G28 ; home all axis
G29 A ; load and activate UBL mesh - alt to M420 S1 L0 Z2 and  M420 S1 Z2
G1 Z50 F240
G1 X2 Y10 F3000
M104 S{first_layer_temperature[0]} ; set final nozzle temp
M190 S{first_layer_bed_temperature[0]} ; wait for bed temp to stabilize
M109 S{first_layer_temperature[0]} ; wait for nozzle temp to stabilize
G1 Z0.28 F240
G92 E0
G1 Y140 E10 F1500 ; prime the nozzle
G1 X2.3 F5000
G92 E0
G1 Y10 E10 F1200 ; prime the nozzle
G92 E0
```

**Reply by DoogH2004** [reply]
Yeah, I am not sure what is going on.  I wish I had a copy of the stock firmware, but the Ender 3 V2 Neo doesn't have the same firmware on their website.

**Reply by mriscoc** [reply]
> 11. Start print and observe the GUI indicator for mesh compensation turning off during `G28` and back on during `G29 A` and same result for `M420 S1 ...` commands in separate tests

If you see the mesh active indicator in the display then the UI correctly informs the firmware core that leveling must be active and the core responds by saying that it has enabled mesh leveling. From there, is the Marlin core who controls how the leveling must be applied without the ProUI intervention.

**Reply by kileywm** [reply]
@mriscoc that makes sense and so far I agree with you. All signs point to the firmware correctly applying a mesh during the print, so I am trusting the UI's indicator that leveling is active.

So why do I have bed leveling issues while I print, despite correct application of my UBL mesh? I have an idea. I'm following a lead now that my bed-leveling issues are more due to the formation of the mesh itself. This could be a hardware issue. Anyway that's beyond the scope of this discussion.

If anyone is curious, I was clued into that lead when I loaded a mesh and started checking mesh points one-by-one for gap consistency. With my S1, stock hardware and Ender3S1-F1-UBL-20220716.bin firmware, the gaps are wildly different, so I have some big problems with mesh generation to sort out.

**Reply by mriscoc** [reply]
Not all S1 have problems with the mesh leveling but there are several reports about that, so, it is not an isolated case. There is a theory about the large offsets that sprite extruder needs and it could introduce inaccuracy.

**Reply by Friday13th87** [reply]
In your start g-code you activate UBL with "G29 A" but you are not loading any mesh.
So what you need is:
G29 A ;activating the mesh
G29 F2.0 ;set fading high,
G29 L0 ;load mesh slot 0
G29 J0 ;3-point mesh tilt

I came to that thread because i searched for the "M420" command, which i never saw and never used and i did not find a clear statement till now.

**Comment by carnaue** [comment]
30% ish of S1 owners have found that the mesh is created and applied but it
does not match the actual build plate surface.  If your problem is that the
filament is too high and low in spots that does not necessarily mean the
mesh is not applied it may just be the calculated mesh values are not
correct.  Use the power of UBL to manually edit your mesh values regardless
of the ABL result, so those areas that are high manually lower them etc.
Follow the instructions linked below that 100's of people have fixed their
own similar problems including myself.

https://www.printables.com/model/213992-perfect-1st-layers-ender-3-v2-ender-3s1-ender-3s1-

start with a 3x3 mesh once that is dialled in you can increase mesh size.

**Reply by Friday13th87** [reply]
If you do it like that, it would be better to just create a manual mash from stack and you dont need a bltouch anymore.
